### DISCUSSION TITLE
Galaxy Watch 4 as collector on WearOS 3.0 problem

### DISCUSSION THREAD

**Discussion by aleks372** [NightscoutFoundation/xDrip#2115] [Q&A]
Hi,

I'm using Watch 4 with latest OneUI 4 (WearOS 3.0).
xDrip latest nightly version as for May 10, 2022.
I have managed to set up Wear Collection service on watch, and it's working!!! But not as it should.

The **main problem** is that watch receives sugar readings **only once** in 10 minutes.

I'm using **Dexcom ONE** sensor (G5/G6 collector).

On watch location service permission set to **always allow**.

From logs as i understand this is caused because on first try wearOB1Collector UNBOUNDS, then waits again 5 minutes and only then tries to authenticate again and succeeds to receive readings.

Is there any way this issue can be resolved in near future or maybe some special tweaks can be added to xDrip for those who use Galaxy Watch 4?

This watch is really powerful with newest WearOS 3.0 and i believe is very popular in Europe at the moment.

![Screenshot_20220511-040322_xDrip+](https://user-images.githubusercontent.com/72258515/167748532-dab62568-8e54-4739-a4b3-100fc6f84e5d.jpg)

**Comment by Navid200** [comment]
@psonnera

**Comment by psonnera** [comment]
I believe this is an historical weakness of Wear OS watches and few were patched to fix that behaviour https://github.com/NightscoutFoundation/xDrip/wiki/Patching-Android-Wear-devices-for-use-with-the-G5 . 10 minutes readings is a good result, most just stop after one.

**Reply by aleks372** [reply]
wear os 3.0 has only recently appeared and there is only a few watches available with it on the market, it has a lot of possibilities.
if only developers updated xDrip for wear os 3 full support.

**Comment by old-square-eyes** [comment]
I got this watch today and I can't get it to collect at all. Location perms enabled.

**Reply by aleks372** [reply]
If you forced watch collection service, it should connect on next update request, try playing with settings.
Also enable wearos logs so you can see what is going on.

**Reply by old-square-eyes** [reply]
I have tried forced. No joy. Been collecting with my SWR50 and Huawei Watch 2 for many years. I know the drill. Still doesn't seem to pick up for some reason.

**Reply by old-square-eyes** [reply]
![image](https://user-images.githubusercontent.com/25760491/168526590-761c054b-06bb-4801-9e96-a65f422b019c.png)

**Reply by old-square-eyes** [reply]
After a few reboots it started collecting. Then it wouldn't sync BG to the phone which I think is this old issue https://github.com/NightscoutFoundation/xDrip/issues/1511 so I faactory reset the watch and set it up from scratch and now it's collecting every 10 mins. Would love to see more Wear support for xDrip but I'm happy for now with this as a backup. Hopefully it works when I come to need it.

**Reply by aleks372** [reply]
First time i tried this, i disconnected phone Bluetooth, so watch connected to phone through cloud over wifi, synced with phone all time on the background.

Buy overall i agree, there is new Wear OS version 3 out, xDrip wear component as i saw from sourcefiles wasnt updated for long now, there could be even a standalone app for watch.

Hope some wear os devs see this and could help xDrip users to catch with latest possibilities 😊

**Comment by jbr7rr** [comment]
Hi,

I ran in to similar problems on the Fossil 5e (wear OS 2.34)
I did some quick hacking here: https://github.com/jbr7rr/xDrip/tree/jbr7rr-wear and got it working. But I'm not sure if it will break other stuff for that I will have to dive deeper into it. For now I have other priorities and may pick this up later.

If you know how to compile the code you can grab the branch and try it on the Galaxy Watch 4. I'm not comfortable releasing an APK or PR yet for this.

**Comment by old-square-eyes** [comment]
Came to need collection at the beach and this failed me. Can't get it to collect :( I had it working prior and mostly use my phone. But really need to be able to collect in the absence of my phone. Any tips on what to try? I really don't want to be haphazardly factory resetting things if I can help it.

**Reply by chachonen** [reply]
Hi, any updates on this? I've followed the instructions but I'm unable to use the watch to collect data. I get the data on the watch (LTE), but it must be coming from the phone, as soon as the phone is out of range from the transmitter I see no data.

I'd really appreciate any guidance or ideas

Thanks 
Henrik

**Comment by old-square-eyes** [comment]
Well. I'm loosing my mind. Starting to doubt it ever worked. Maybe it failed back to the phone for a couple of readings. Dunno. I was pretty sure it was because it was collecting every 10mins... the hallmark of crappy Wear collection.

**Reply by UberQueiroz** [reply]
For me not only my GW4 can't collect the sensor's readings but also it isn't automatically unabling force collector after X minutes of missed readings.

**Comment by WanaGo** [comment]
Curious if there has been any updates to this whole Collecting situation? 

I got a **Samsung Galaxy Watch 5 Pro LTE** for my 40th (which uses the Google Wear OS 3.5), and sideloaded xDrip with the Wear Installer 2 app from my phone, so that side of things is working great. Get the usual watch face that I have had for years on my old watch (Sony Smartwatch 3), and get alarms etc on the watch, and can snooze them no problem.

But I just cant get collector to work, not even once every 10 mins etc.
Really could have used it the other day, as phone was about to die, but it just wouldnt collect, no matter what I tried.
Would be really neat if there is a work around for this.

Thanks

**Reply by Navid200** [reply]
There are not many developers who work on wear.  We need developers.

**Reply by jbr7rr** [reply]
I do have something coded up that is a bit hacky but works on my galaxy watch 4. I can try to nicen it up and get a PR in but before I spent the time on that I would like to know:
- Are there ppl who are going to use it, that can also test it on different (new and old) wear devices?
- Do the xDrip+ devs support this effort? That is, once I made a PR does it have a chance of being merged?

**Reply by old-square-eyes** [reply]
I would use it and happily test it

**Reply by old-square-eyes** [reply]
Galaxy Watch 4

**Reply by WanaGo** [reply]
Happy to Test and Use it also, please do share :)

**Reply by rudysroost** [reply]
I would test on my Pixel watch. The only reason I've bought my 3 watches is to use them as a collector. So far, it hasn't worked on any. I had high hopes for the Pixel watch.

**Reply by TheCassos** [reply]
same here, suspect many many more
hoping to unlock some more freedom for my 8 yr old
- have a Galaxy Watch 4 classic LTE here, if you want a tester

**Reply by UberQueiroz** [reply]
I have a GW4 (WearOS 3) and also a TWP3 (WearOS 2). I would be glad to test it on both watches.

**Reply by TheCassos** [reply]
@jamorham

**Reply by aleks372** [reply]
@jbr7rr that's awesome, we definetly need some movement  on this collection issue to make Wear 3.0 collection working.
Please give us more info when you ready, i would be happy to test it on my Watch4.

**Comment by chachonen** [comment]
Same here, happy to test and use it!

**Comment by Navid200** [comment]
@jbr7rr 

The following is my opinion.  I am not making an official statement on behalf of the developers.

There is a lot of redundant code in xDrip because of Wear.
Many active xDrip developers do not use a watch.  Or, they don't use WearOS.  So, when they make a tweak to xDrip, they don't update the corresponding WearOS code.  The problem is that even if they do, there is no way to test that part.  So, I don't even think it is a good idea to update that code if the person updating it is not going to test it.

I have always wondered if we can move the wearOS portion out of xDrip itself.
As an example, please look at how Miband is managed.  There is a service in xDrip that sends raw data to a companion app (WatchDrip).  It's the companion app that communicates with the watch.
The code in the companion app is managed by a different individual outside the xDrip project.  He happens to be one of the xDrip developers.
This allows him to do what he needs with no need to wait for this repository to approve his changes.
May be once a year, he may need to tweak the service part of xDrip.  But, overall, the number of PRs he needs to get approval for in this repository have significantly decreased since Miband moved out of xDrip.
When a new Miband comes to the market, all he needs to do is to modify the companion app.

You may want to consider a disadvantage of this approach the requirement to run xDrip as well as a companion app on the phone.
But, in my opinion, the advantages significantly overshadow this disadvantage.

There are many individuals who love to use wear OS.  There are many xDrip users who don't.  The current situation of not having any active developer who updates xDrip with respect to wear has resulted in the wear not being up-to-date in xDrip.

Are you interested in doing something similar for Wear?
So, a service will need to transfer only raw values to the companion app.  The companion app needs to do all the analysis that xDrip used to do in the redundant code to calculate the resulting parameters and drive the wear watches.
If this succeeds, we will be able to remove all the wearOS redundant code from xDrip.

I cannot emphasize enough that this is not an official message from @jamorham.   This is only my opinion.

**Reply by jbr7rr** [reply]
While I agree that conceptional splitting of wear development is a good idea, I think that practically it's a lot of effort and difficult to do properly, as lots of wear functions and settings are really intertwined with the main app. There is way more functionality then just getting the CGM reading to the watch. For example the collector hand over between app and wear. 

On the other hand updating and fixing the wear code to be in line with the current app state is also a lot of work. especially if we go further then only the collector part. But it's less complex and adds ease of use for users as they don't have to use a separate app and can install to wear more easily (but like you said this is only a minor detail/inconvenience).

Also backwards compatibility is also an issue, As I can only test on Wear OS2.5 and Wear OS 3.5. Not on the older Wear OS 1.x versions where collector is working? But that will be an issue regardless splitting of the wear component or not.

All thoughts on this are welcome, as we can see there is a user need/want for this. But I also want to keep it maintainable and doable for everyone involved.

**Reply by chachonen** [reply]
Hi, my use case is just remote monitoring the glucose levels without my daughter having to take a mobile phone with her. I'm no mobile developer, but, would the following approach be viable? Have a service on the watch collect data (probably extracting the needed modules from the xdrip app + handling which device gets to read the data from the sensor + all other obstacles that will come up), then have the service forward data to xdrip (maybe a companion app on the phone, maybe other data sources...no idea if this is possible). And have xdrip do it's thing on notifying followers. I know there is a lot of useful stuff left out, but that is my primary need.

Thanks!
Henrik

**Comment by old-square-eyes** [comment]
I can donate a SWR50 to the cause and a Huawei Smart Watch. Two of the most commonly used older watches, for backwards compatibility tests.

Also... Have the main devs reviewed the hack already posted? Maybe that will be fine until it can be done properly?

**Comment by aalaamer** [comment]
> Hi,
> 
> I'm using Watch 4 with latest OneUI 4 (WearOS 3.0). xDrip latest nightly version as for May 10, 2022. I have managed to set up Wear Collection service on watch, and it's working!!! But not as it should.
> 
> The **main problem** is that watch receives sugar readings **only once** in 10 minutes.
> 
> I'm using **Dexcom ONE** sensor (G5/G6 collector).
> 
> On watch location service permission set to **always allow**.
> 
> From logs as i understand this is caused because on first try wearOB1Collector UNBOUNDS, then waits again 5 minutes and only then tries to authenticate again and succeeds to receive readings.
> 
> Is there any way this issue can be resolved in near future or maybe some special tweaks can be added to xDrip for those who use Galaxy Watch 4?
> 
> This watch is really powerful with newest WearOS 3.0 and i believe is very popular in Europe at the moment.
> 
> ![Screenshot_20220511-040322_xDrip+](https://user-images.githubusercontent.com/72258515/167748532-dab62568-8e54-4739-a4b3-100fc6f84e5d.jpg)

Can you tell me how you did it

**Comment by old-square-eyes** [comment]
Amazing it started working for me too. Must be the new WearOS 4 which just came out.

Only issue is I get the no sync back to the phone issue I reported earlier. Can't have em all.

Mine also every 10 mins.

**Reply by Unai-HP** [reply]
What are your settings? Mine does not work even in Wear4. Thanks

**Reply by old-square-eyes** [reply]
It worked for a few readings then stopped. Hasn't worked again since in trying several times. No special settings.
